tests/backends/postgresql/tests.py

```diff
@@ -223,6 +223,82 @@ def test_connect_non_autocommit(self):
         finally:
             new_connection.close()
 
+    @unittest.skipUnless(is_psycopg3, "psycopg3 specific test")
+    def test_connect_pool(self):
+        from psycopg_pool import PoolTimeout
+
+        new_connection = connection.copy()
+        self.assertIsNone(new_connection.pool)
+
+        new_connection.settings_dict["OPTIONS"]["pool"] = {
+            "min_size": 2,
+            "timeout": 0.1,
+        }
+        new_connection._connection_pools = {}
```

Comment:
This line clears all of the process-level connection pools without closing them, which might trigger some ResourceWarnings in the Django test suite, in case you care about those.

I don't know if the DJango test suite has the notion of running specific tests in isolated processes. If it does, then this is a good candidate for it.

@sarahboyce I'm curious if you tried to run the Django test suite itself with a connection pool configuration. It will be a good exercise to see what breaks if you do this (ideally everything passes...).

Reply:
I have tweaked the code and improved the after test cleanup (maybe check again and see if you think that's better?)

> I'm curious if you tried to run the Django test suite itself with a connection pool configuration.
I hadn't but that is a good idea. It initially failed immediately with:
```
Found 16748 test(s).
Creating test database for alias 'default'...
Cloning test database for alias 'default'...
Got an error creating the test database: source database "test_django" is being accessed by other users
DETAIL:  There are 4 other sessions using the database.
```
Which I was able to get around by setting `parallel` to 1, but I think that means there is an issue with `_clone_test_db` which tries to close connections (and maybe needs to close the pool? Not sure yet)
Then 16k tests not parallized takes a little time, looks like a handful of failures (a couple are expected based off the way I wrote the pool tests) but mostly passing.
Going to try to look into `_clone_test_db`

Reply:
4244d71d829a717a1a871d52bdda910d173bb6b1 this means I can now run them in parallel
There's quite a lot to look into here, mostly timezone tests are failing. But this is a simplified list of the failing tests
```
db_functions.datetime.test_extract_trunc.DateFunctionWithTimeZoneTests
timezones.tests.LegacyDatabaseTests
timezones.tests.NewFormsTests.test_model_form
cache.tests.DBCacheWithTimeZoneTests

backends.postgresql.tests.Tests.test_connect_and_rollback
backends.postgresql.tests.Tests.test_connect_custom_cursor_factory
backends.postgresql.tests.Tests.test_connect_role
backends.tests.BackendTestCase.test_is_usable_after_database_disconnects
```

Reply:
80f11f5ec3938bad12a97b6be447fb65ca63ef7b this takes care of the timezone tests and then only 
Only `backends.postgresql.tests.Tests.test_connect_custom_cursor_factory` I think is maybe worth looking into now and I think that is due to the settings changes during the test itself
So basically there

Reply:
Hmm regarding the change in 80f11f5: removing the `and not self.pool` condition from `init_connection_state` is odd because it means that `_configure_connection` is now runs twice, once by the pool's `configure` callback, and once by `init_connection_state`. Were you able to understand why it's needed?

Reply:
So by not doing this for pool, the timezone changes were never commited and then broke tests when using a pool setup.
I have now removed the logic refactor creating `_configure_connection`.

Reply:
I am not sure that the new refactor is correct. Now we configure the connection every time on pool checkout this is certainly wrong! FWIW if the tests fail with the pool it might be that they also fail currently with persistent connections.

Reply:
Let me expand on my previous comment to explain why I think that the latest refactor is wrong:

* With default connections and non-persistent connections we close a connection every request and reopen it every request. As such we need one query to set the timezone every request
* With persistent connections the connection stays open and the timezone is only set once per connection. The assumption is that the view doesn't change the timezone in a bad way.
* With a pool we are basically back to the first case, as far as Django is concerned we are closing the connection every request (but in reality put it back to the pool) and once we check it out again it should not net to apply a new timezone (ie it behaves like a persistent connection)

To test the behavior I would recommend the following: 

 * Disable pooling, set a timezone in the settings that is != your default timezone and enable query logging in postgres. Verify that the timezone is set every request.
 * Now enable persistent connections and verify that the timezone is only set once per connection
 * Now disable persistent connections and enable a pool with a maxsize of 1, verify that the timezone is set once per connection

Reply:
I investigated and `cache.tests.DBCacheWithTimeZoneTests` was failing because it was running `createcachetable` management command on test setUp (which then creates a database connection in the pool). If I add similar logic to `_close_all_connections` which I am running on on the test database cloning/creating methods, then this all passes.

I think these issues are more around test state clean up which I guess I'm still trying to figure out.

I can add tests around the timezone setting so I don't remove the logic again 😁